crates/rpc/rpc-builder/src/lib.rs

```diff
@@ -1710,39 +1724,75 @@ impl RpcServerConfig {
         let mut ws_server = None;
         if let Some(builder) = self.ws_server_config.take() {
             let builder = builder.ws_only();
-            let (server, addr) = WsHttpServerKind::build(
-                builder,
-                ws_socket_addr,
-                self.ws_cors_domains.take(),
-                self.jwt_secret.clone(),
-                ServerKind::WS(ws_socket_addr),
-                modules.ws.as_ref().map(RpcRequestMetrics::ws).unwrap_or_default(),
-            )
-            .await?;
+
+            let cors_layer: Option<CorsLayer> = self
+                .ws_cors_domains
+                .as_deref()
+                .map(cors::create_cors_layer)
+                .and_then(|res| res.ok())
+                .or_else(|| None);
+            let jwt_auth: Option<AuthLayer<JwtAuthValidator>> =
+                jwt_secret.map(|secret| AuthLayer::new(JwtAuthValidator::new(secret)));
+            let middleware =
+                tower::ServiceBuilder::new().option_layer(cors_layer).option_layer(jwt_auth);
+
+            let server = builder
+                .set_http_middleware(middleware)
+                .set_rpc_middleware(
+                    RpcServiceBuilder::new()
+                        .layer(modules.ws.as_ref().map(RpcRequestMetrics::ws).unwrap_or_default()),
+                )
+                .build(ws_socket_addr)
+                .await
+                .map_err(|err| RpcError::server_error(err, ServerKind::WS(ws_socket_addr)))?;
+            let addr = server
+                .local_addr()
+                .map_err(|err| RpcError::server_error(err, ServerKind::WS(ws_socket_addr)))?;
+
             ws_local_addr = Some(addr);
             ws_server = Some(server);
         }
 
         if let Some(builder) = self.http_server_config.take() {
             let builder = builder.http_only();
-            let (server, addr) = WsHttpServerKind::build(
-                builder,
-                http_socket_addr,
-                self.http_cors_domains.take(),
-                self.jwt_secret.clone(),
-                ServerKind::Http(http_socket_addr),
-                modules.http.as_ref().map(RpcRequestMetrics::http).unwrap_or_default(),
-            )
-            .await?;
-            http_local_addr = Some(addr);
+
+            let cors_layer: Option<CorsLayer> = self
+                .http_cors_domains
+                .take()
+                .as_deref()
+                .map(cors::create_cors_layer)
+                .and_then(|res| res.ok())
+                .or_else(|| None);
+            let jwt_auth: Option<AuthLayer<JwtAuthValidator>> = self
+                .jwt_secret
+                .as_ref()
+                .map(|secret| AuthLayer::new(JwtAuthValidator::new(secret.clone())));
```

Comment:
i'm seeing this cors/jwt logic almost 1:1 duplicated 3 times here, let's try to unify?